### ISSUE TITLE
ar_pose + ardrone_autonomy

### ISSUE THREAD

**Issue by pulver22** [AutonomyLab/ardrone_autonomy#181]
Hi all,

I'm trying to perform marker detection using the`ar_pose` package. 
Everything works fine using the package alone, either wih the webcam or an usb camera, getting the video stream with gscam.

I tried to write a launch file that launch `ar_pose` and `ardrone_autonomy` simultaneously, in order to use the frontcam of the quadcopter as camera for the acquisition. Here is the launch file:

``` xml
<launch>
    <!-- IPv4 address of your drone -->
    <arg name="ip" default="192.168.1.1"/>

    <!-- Ultrasound frequency (7 or 8). -->
    <arg name="freq" default="8"/>
    <node name="ardrone_driver" pkg="ardrone_autonomy" type="ardrone_driver" output="screen" clear_params="true" args="-ip $(arg ip)">
        <param name="outdoor" value="0"/>
        <param name="max_bitrate" value="4000"/>
        <param name="bitrate" value="4000"/>
        <param name="navdata_demo" value="0"/>
        <param name="flight_without_shell" value="0"/>
        <param name="altitude_max" value="4000"/>
        <param name="altitude_min" value="50"/>
        <param name="euler_angle_max" value="0.21"/>
        <param name="control_vz_max" value="700"/>
        <param name="control_yaw" value="1.75"/>
        <param name="detect_type" value="10"/>
        <param name="enemy_colors" value="3"/>
        <param name="detections_select_h" value="32"/>
        <param name="detections_select_v_hsync" value="128"/>
        <param name="enemy_without_shell" value="0"/>
        <param name="ultrasound_freq" value="$(arg freq)"/>
        <param name="realtime_navdata" value="true"/>
        <param name="realtime_video" value="true"/>
        <!-- Covariance Values (3x3 matrices reshaped to 1x9)-->
        <rosparam param="cov/imu_la">[0.1, 0.0, 0.0, 0.0, 0.1, 0.0, 0.0, 0.0, 0.1]</rosparam>
        <rosparam param="cov/imu_av">[1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]</rosparam>
        <rosparam param="cov/imu_or">[1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 100000.0]</rosparam>
    </node>

    <!-- Launch the static transform publisher for world to camera -->
    <!--<node pkg="tf" type="static_transform_publisher" name="world_to_cam" args="0 0 0.5 -1.57 0 -1.57 world camera 10"/>
    <node pkg="tf" type="static_transform_publisher" name="base_to_frontcam" args="0.21 0 0 -1.57 0 -1.57 ardrone_base_link ardrone_base_frontcam 10"/>
    <node pkg="tf" type="static_transform_publisher" name="base_to_bottomcam" args="0 -0.02 0.0 3.14 0 1.57 ardrone_base_link ardrone_base_bottomcam 10"/>-->
    <!--<node pkg="tf" type="static_transform_publisher" name="cam_to_frontcam" args="0.5 0 0 0 0 0 camera ardrone_base_frontcam 10"/>-->
    <!--<node pkg="tf" type="static_transform_publisher" name="cam_to_dronecam" args="0.2 0 0 0 0 0 camera ardrone_base_frontcam 10"/>-->

    <node pkg="tf" type="static_transform_publisher" name="cam_to_drone" args="1 0 0 0 0 0 world ardrone_base_frontcam 10"/>

    <!-- Launch the image rectification node -->
    <node ns="camera" pkg="image_proc" type="image_proc" name="image_proc"/>

    <!-- Start the GSCAM node -->
    <!--<env name="GSCAM_CONFIG" value="v4l2src device=/dev/video0 ! video/x-raw-yuv,framerate=30/1,width=640,height=480 ! ffmpegcolorspace "/>
    <node pkg="gscam" type="gscam" name="gscam" output="screen">
        <param name="width" type="int" value="640"/>
        <param name="height" type="int" value="480"/>
        <param name="fps" type="int" value="30"/>
        <param name="frame_id" type="string" value="ardrone_base_frontcam"/>
        <param name="camera_info_url" type="string" value="file://$(find gscam)/camera_calibration.yaml"/>
    </node>-->

    <node name="ar_pose" pkg="ar_pose" type="ar_single" respawn="false" output="screen">
        <param name="marker_pattern" type="string" value="$(find ar_pose)/data/4x4/4x4_1.patt"/>
        <param name="marker_width" type="double" value="152.4"/>
        <param name="marker_center_x" type="double" value="0.0"/>
        <param name="marker_center_y" type="double" value="0.0"/>
        <param name="threshold" type="int" value="100"/>
        <param name="use_history" type="bool" value="true"/>
    </node>

    <!-- Launch RVIZ as visualizer -->
    <!--<node pkg="rviz" type="rviz" name="rviz" args="-d $(find ar_pose)/launch/live_single.rviz"/>-->

</launch>
```

The problem is that in this way no marker are detected. 
If I uncomment the gscam part of the launch file, I am able to perceive marker only from `device=/dev/video0` that is the webcam.

Have you ever tried to perform this operation?

I think one solution could be modified the frame of ardrone in the source, changing `ardrone_base_frontcam` in `camera` but I'm not sure. I think so because it seems that, as `ar_pose` is written, marker can be attached only to `camera` frame (of course another attempt woulbe changing the `ar_pose` code).

in the hypothesis I don't want to change the code (to not recompile it) is there a way to do it inside the launch file?

**Comment by mani-monaj** [comment]
I think you need a proper [remapping](http://wiki.ros.org/roslaunch/XML/remap) from `ardrone/image_raw` to `camera/image_raw` (or whatever topic `arpose` requires as an input). If you paste the result of `rostopic list` on your system, I can suggest the correct `<remap>` tag.

**Comment by pulver22** [comment]
Here's the output of `rostopic list`:

```
/ar_pose_marker
/ardrone/bottom/camera_info
/ardrone/bottom/image_raw
/ardrone/bottom/image_raw/compressed
/ardrone/bottom/image_raw/compressed/parameter_descriptions
/ardrone/bottom/image_raw/compressed/parameter_updates
/ardrone/bottom/image_raw/compressedDepth
/ardrone/bottom/image_raw/compressedDepth/parameter_descriptions
/ardrone/bottom/image_raw/compressedDepth/parameter_updates
/ardrone/bottom/image_raw/theora
/ardrone/bottom/image_raw/theora/parameter_descriptions
/ardrone/bottom/image_raw/theora/parameter_updates
/ardrone/camera_info
/ardrone/front/camera_info
/ardrone/front/image_raw
/ardrone/front/image_raw/compressed
/ardrone/front/image_raw/compressed/parameter_descriptions
/ardrone/front/image_raw/compressed/parameter_updates
/ardrone/front/image_raw/compressedDepth
/ardrone/front/image_raw/compressedDepth/parameter_descriptions
/ardrone/front/image_raw/compressedDepth/parameter_updates
/ardrone/front/image_raw/theora
/ardrone/front/image_raw/theora/parameter_descriptions
/ardrone/front/image_raw/theora/parameter_updates
/ardrone/image_raw
/ardrone/image_raw/compressed
/ardrone/image_raw/compressed/parameter_descriptions
/ardrone/image_raw/compressed/parameter_updates
/ardrone/image_raw/compressedDepth
/ardrone/image_raw/compressedDepth/parameter_descriptions
/ardrone/image_raw/compressedDepth/parameter_updates
/ardrone/image_raw/theora
/ardrone/image_raw/theora/parameter_descriptions
/ardrone/image_raw/theora/parameter_updates
/ardrone/imu
/ardrone/land
/ardrone/mag
/ardrone/navdata
/ardrone/odometry
/ardrone/reset
/ardrone/takeoff
/camera/camera_info
/camera/image_color
/camera/image_color/compressed
/camera/image_color/compressed/parameter_descriptions
/camera/image_color/compressed/parameter_updates
/camera/image_color/compressedDepth
/camera/image_color/compressedDepth/parameter_descriptions
/camera/image_color/compressedDepth/parameter_updates
/camera/image_color/theora
/camera/image_color/theora/parameter_descriptions
/camera/image_color/theora/parameter_updates
/camera/image_mono
/camera/image_mono/compressed
/camera/image_mono/compressed/parameter_descriptions
/camera/image_mono/compressed/parameter_updates
/camera/image_mono/compressedDepth
/camera/image_mono/compressedDepth/parameter_descriptions
/camera/image_mono/compressedDepth/parameter_updates
/camera/image_mono/theora
/camera/image_mono/theora/parameter_descriptions
/camera/image_mono/theora/parameter_updates
/camera/image_proc_debayer/parameter_descriptions
/camera/image_proc_debayer/parameter_updates
/camera/image_proc_rectify_color/parameter_descriptions
/camera/image_proc_rectify_color/parameter_updates
/camera/image_proc_rectify_mono/parameter_descriptions
/camera/image_proc_rectify_mono/parameter_updates
/camera/image_rect
/camera/image_rect/compressed
/camera/image_rect/compressed/parameter_descriptions
/camera/image_rect/compressed/parameter_updates
/camera/image_rect/compressedDepth
/camera/image_rect/compressedDepth/parameter_descriptions
/camera/image_rect/compressedDepth/parameter_updates
/camera/image_rect/theora
/camera/image_rect/theora/parameter_descriptions
/camera/image_rect/theora/parameter_updates
/camera/image_rect_color
/camera/image_rect_color/compressed
/camera/image_rect_color/compressed/parameter_descriptions
/camera/image_rect_color/compressed/parameter_updates
/camera/image_rect_color/compressedDepth
/camera/image_rect_color/compressedDepth/parameter_descriptions
/camera/image_rect_color/compressedDepth/parameter_updates
/camera/image_rect_color/theora
/camera/image_rect_color/theora/parameter_descriptions
/camera/image_rect_color/theora/parameter_updates
/cmd_vel
/rosout
/rosout_agg
/tf
/visualization_marker
```

I already tried to remap the topic but it didn't work properly. Maybe I did it in the wrong way!

**Comment by mani-monaj** [comment]
Try this:

``` xml
<node name="ardrone_driver" pkg="ardrone_autonomy" type="ardrone_driver" output="screen" clear_params="true" args="-ip $(arg ip)">
    ...
    <remap from="ardrone/front/image_raw" to="camera/image_color" />
    ...
</node>
```

Also, you probably do not need the `static_transform_publisher` node since the driver already [publishes](http://ardrone-autonomy.readthedocs.org/en/latest/frames.html) the proper transform from the camera frame to the `base_link`.

**Comment by pulver22** [comment]
The problems are partially solved. 

Now I can detect the marker but its frame is fixed to the one of the drone camera [link](http://imgur.com/NzyQBpem9CraZV) instead moving around as it should do [link](http://imgur.com/m9CraZV).

**Comment by mani-monaj** [comment]
Can you please visualize all TF frames for both of the cases and share the results (`frame.pdf`):

``` bash
$ rosrun tf view_frames
```

**Comment by pulver22** [comment]
I solved the problem, I'll try to clarify the situation.

I discovered that the marker was attached to the camera frame because I was visualizing the Image received on the topic, without its depth. 

Trying to visualize the `Camera` instead of the `Image` in Rviz I got the following error:

`Error: CameraInfo/P resulted in an invalid position calculation`

Investigating the error above I discovered that the camera need to be calibrated to perceive the depth of the image broadcasted on the topic.

After doing so I was able to see the marker moving around the camera frame!!

Thank you @mani-monaj for your support!
